### PR TITLE
Make user able to change all color palette origins

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
This PR makes the user able to edit default, theme, and custom color palettes. For the default and theme the user can only change color values.
For the custom palette, the user can change everything.
The user can reset each palette individually.

It also prefixes also custom colors slugs with "custom-".

### CODE REVIEW

**Comment by oandregal** [comment]
Hey, I've tested this and it's working nicely. Need to take a look at the code (will do it today before lunch).

There's only one thing we should fix before landing: the user can introduce spaces (and other characters) in the custom color slug, so we should sanitize this (no spaces, valid characters), etc.

Other things/follow-ups/up for discussion:

- UI: can the color values be editable by clicking directly in the color circle? It wasn't obvious to me that I should click the three dots in the top-right to edit them.
- I tried to add a different palette to a block (paragraph) and it works fine. Then, I tried to make that palette empty but I couldn't (the global colors are used when I do that). Should users be able to "empty" the palette for a context different than the global one?
- The use of suffix `custom`. What do you think of separating the `custom` infix with `--`? This is, the CSS variable would be `--wp--preset--color--custom--<color-value>` instead of the current `--wp--preset--color--custom-<color-value>`?

File: lib/class-wp-theme-json-gutenberg.php [hunk 1]
```diff
@@ -1333,48 +1335,52 @@ public function merge( $incoming ) {
 	 * @return array
 	 */
 	private static function remove_insecure_settings( $input ) {
+		$origins = array( 'default', 'theme', 'user' );
```
Comment (oandregal) [inline lib/class-wp-theme-json-gutenberg.php]: Can we use the `VALID_ORIGINS` here?

File: lib/class-wp-theme-json-gutenberg.php [hunk 2]
```diff
@@ -301,7 +301,9 @@ public function __construct( $theme_json = array(), $origin = 'theme' ) {
 				$path   = array_merge( $node['path'], $preset_metadata['path'] );
 				$preset = _wp_array_get( $this->theme_json, $path, null );
 				if ( null !== $preset ) {
-					_wp_array_set( $this->theme_json, $path, array( $origin => $preset ) );
+					if ( 'user' !== $origin || isset( $preset[0] ) ) {
```
Comment (oandregal) [inline lib/class-wp-theme-json-gutenberg.php:304]: I'd think we want to rename the `user` origin to `custom`, as we did with `core` to `default` to match the UI, class/css var generation, etc. Perhaps we can do this in a different PR that lands before this one?

Comment (oandregal) [reply lib/class-wp-theme-json-gutenberg.php:304]: What's the use case `isset( $preset[ 0 ] )` is checking for? My understanding is that we should always skip reorganizing the presets if the origin is the user, as we'll store them already keyed by origin, so this could be simplified to: `if ( null != $preset && 'user' !== $origin )`.

Additionally, I wonder if we need to keep back compatibility with existing user data in the database. We could add an extra step to normalize the user data by origin for that case (older data that is not keyed by origin).

File: packages/components/src/color-edit/index.js [hunk 3]
```diff
@@ -213,24 +227,26 @@ export default function ColorEdit( { colors = EMPTY_ARRAY, onChange } ) {
...
-							setIsEditing( true );
-							setEditingColor( colors.length );
-						} }
-					/>
+					{ ! canOnlyChangeValues && (
+						<Button
+							isSmall
+							isPressed={ isAdding }
+							icon={ plus }
+							label={ __( 'Add custom color' ) }
```
Comment (oandregal) [inline packages/components/src/color-edit/index.js]: It sounds like the ColorEdit component should be agnostic as to what's the source of the color. I'd think this label should be provided by the consumer or be called `Add color`.

File: packages/components/src/color-edit/index.js [hunk 4]
```diff
@@ -68,22 +70,22 @@ function ColorOption( {
...
-					{ isEditing ? (
+					{ isEditing && ! canOnlyChangeValues ? (
 						<ColorNameInput
 							value={ color.name }
 							onChange={ ( nextName ) =>
 								onChange( {
 									...color,
 									name: nextName,
-									slug: kebabCase( nextName ),
+									slug: slugPrefix + kebabCase( nextName ),
```
Comment (oandregal) [inline packages/components/src/color-edit/index.js:80]: Ah, I see. We need to fix the kebabCase import to `import { kebabCase } from 'lodash';`.

File: packages/components/src/color-edit/index.js [hunk 5]
```diff
@@ -49,12 +49,14 @@ function ColorNameInput( { value, onChange } ) {
...
 
 function ColorOption( {
+	canOnlyChangeValues,
 	color,
 	onChange,
 	isEditing,
 	onStartEditing,
 	onRemove,
 	onStopEditing,
+	slugPrefix,
```
Comment (oandregal) [inline packages/components/src/color-edit/index.js:59]: I'm not very keen on making the ColorEdit component aware of the prefix for custom colors. For components, do we usually expose "hooks" for consumers to tweak the normal behavior? Like, for example, we could have an `onNameChange` callback that consumers can use to tweak the name as they like (add a prefix, etc).

Not really familiar with how/whether this is a practice we want our components to have, so I'll go with whatever you think is best.

File: packages/edit-site/src/components/global-styles/color-palette-panel.js [hunk 6]
```diff
@@ -1,22 +1,72 @@
...
+			{ !! themeColors && !! themeColors.length && (
+				<ColorEdit
+					canReset={ themeColors !== baseThemeColors }
+					canOnlyChangeValues
+					colors={ themeColors }
+					onChange={ setThemeColors }
+					paletteLabel={ __( 'Theme' ) }
+				/>
+			) }
+			{ !! defaultColors && !! defaultColors.length && (
```
Comment (oandregal) [inline packages/edit-site/src/components/global-styles/color-palette-panel.js]: We also need to consider themes that opt-out from the default palette via `settings.color.defaultPalette`.

**Comment by jorgefilipecosta** [comment]
Hi @oandregal,

> There's only one thing we should fix before landing: the user can introduce spaces (and other characters) in the custom color slug, so we should sanitize this (no spaces, valid characters), etc.


I will try to address that but I think it should be in another PR that is an issue we already had and not related to the changes we are doing here.


> UI: can the color values be editable by clicking directly in the color circle? It wasn't obvious to me that I should click the three dots in the top-right to edit them.

No, one can not directly edit color by pressing the color circle. We are keeping the same behavior as before. In the mockups, I don't see colors being edited by pressing the circle so I don't think that is expected.

> The use of suffix custom. What do you think of separating the custom infix with --? This is, the CSS variable would be --wp--preset--color--custom--<color-value> instead of the current --wp--preset--color--custom-<color-value>


Custom is part of the slug and we use "--" to separate paths in the preset, so in some parts of the engine may be "custom--red" would not be interpreted as a slug. Also, the classes would have strange names like "custom--red-background-color". I would prefer to keep the prefix "custom-", as using "custom--" breaks the promise that "--" is a separator mark in the preset path, (the slug itself would have that mark).

File: packages/components/src/color-edit/index.js (hunk 5 above)
Comment (jorgefilipecosta) [reply packages/components/src/color-edit/index.js:59]: I think the prefix may end up being used for other things besides custom colors, and it seems we will probably not need other name manipulation than the prefix. So I guess slugPrefix is enough. In the future, I'm not totally sure of the API and the place where this component will be. I guess when we remove experimental from this component or move the component to another package we should audit its API.

File: lib/class-wp-theme-json-gutenberg.php (hunk 2 above)
Comment (oandregal) [reply lib/class-wp-theme-json-gutenberg.php:304]: Oh, I see now: the `isset( $preset[ 0 ] )` already does the back-compatibility trick :) Only old user data with no keys by origin would have the `0` key.

**Review by oandregal — APPROVED** [review]
